### PR TITLE
Add efficient hashing of CartesianIndex objects

### PR DESCRIPTION
**Pull request by timholy**
I'm committing this on behalf of one of our GSoC students, @annimesh2809, for whom a source build seems to be crashing.

### CODE REVIEW

File: base/multidimensional.jl [hunk 1]
```diff
@@ -265,6 +275,7 @@ module IteratorsMD
         istop,  jstop  = split(last(R), V)
         CartesianRange(istart, istop), CartesianRange(jstart, jstop)
     end
+
```
Comment (tkelman) [inline base/multidimensional.jl]: unrelated / unnecessary ?

**Review by Sacha0 — APPROVED** [review]
superficially lgtm! :)

File: base/multidimensional.jl [hunk 2]
```diff
@@ -123,6 +123,16 @@ module IteratorsMD
     _isless(ret, ::Tuple{}, ::Tuple{}) = ifelse(ret==1, true, false)
     icmp(a, b) = ifelse(isless(a,b), 1, ifelse(a==b, 0, -1))
 
+    # hashing
+    const cartindexhash_seed = UInt == UInt64 ? 0xd60ca92f8284b8b0 : 0xf2ea7c2e
+    function Base.hash(ci::CartesianIndex, h::UInt)
+        r = h + cartindexhash_seed
```
Comment (nalimilan) [inline base/multidimensional.jl]: Why not reuse `h` instead of introducing `r`?

**Comment by timholy** [comment]
Test failures seem unrelated. But is it normal to have two spurious failures on Travis? Might there be something wrong?

**Comment by martinholters** [comment]
If I'm not mistaken, the dependencies got recompiled. In that case, a timeout is to be expected.

**Pull request merged**